pkg/endpoint/events.go

```diff
@@ -119,6 +120,77 @@ func (e *Endpoint) PolicyRevisionBumpEvent(rev uint64) {
 	}
 }
 
+/// EndpointNoTrackEvent contains all fields necessary to update the NOTRACK rules.
+type EndpointNoTrackEvent struct {
+	ep     *Endpoint
+	annoCB AnnotationsResolverCB
+}
+
+// Handle handles the NOTRACK rule update.
+func (ev *EndpointNoTrackEvent) Handle(res chan interface{}) {
+	e := ev.ep
+	port, err := ev.annoCB(e.K8sNamespace, e.K8sPodName)
+	if err != nil {
+		res <- &EndpointRegenerationResult{
+			err: err,
+		}
+		return
+	}
+	// Validate annotation before we do any actual alteration to the endpoint.
+	_, err = strconv.ParseUint(port, 10, 16)
+	if err != nil {
+		res <- &EndpointRegenerationResult{
+			err: err,
+		}
+		return
+	}
+
+	// If this endpoint is going away, remove all related NOTRACK rules.
```

Comment:
What is going to happen when a DNS pod gets removed while cilium-agent was down? Who will remove its NOTRACK rules?

Reply:
The following code will call `d.deleteEndpointQuiet(...)` -> `ep.Delete(...)`:

https://github.com/cilium/cilium/blob/c140bccad6a5552b37ea1c1901381400c0b02156/daemon/cmd/state.go#L300-L313